File: docs/self-test-vectors/crypto_a256gcm_kat.py
 # SPDX-FileCopyrightText: 2026 Digg - Agency for Digital Government
 #
 # SPDX-License-Identifier: EUPL-1.2
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.exceptions import InvalidTag

def b64u(b):   return base64.urlsafe_b64encode(b).rstrip(b"=").decode("ascii")
def unb64u(s): return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))

# RFC 7516 A.1, transcribed from the RFC's JSON array notation.
plaintext = bytes([84,104,101,32,116,114,117,101,32,115,105,103,110,32,111,102,32,105,110,
 116,101,108,108,105,103,101,110,99,101,32,105,115,32,110,111,116,32,107,110,111,119,108,
 101,100,103,101,32,98,117,116,32,105,109,97,103,105,110,97,116,105,111,110,46])          # A.1
cek = bytes([177,161,244,128,84,143,225,115,63,180,3,255,107,154,212,246,138,7,110,91,112,
 46,34,105,47,130,203,46,122,234,64,252])                                                 # A.1.2
iv  = bytes([227,197,117,252,2,219,233,68,180,225,77,219])                                # A.1.4
rfc_ct  = bytes([229,236,166,241,53,191,115,196,174,43,73,109,39,122,233,96,140,206,120,52,
 51,237,48,11,190,219,186,80,111,104,50,142,47,167,59,61,181,127,196,21,40,82,242,32,123,
 143,168,226,73,216,176,144,138,247,106,60,16,205,160,109,64,63,192])                     # A.1.6
rfc_tag = bytes([92,80,104,49,133,25,161,215,173,101,219,211,136,91,210,145])             # A.1.6

aesgcm = AESGCM(cek)

# 0. Reproduce A.1 exactly. Catches any transcription slip above before it can propagate.
rsa_hdr = "eyJhbGciOiJSU0EtT0FFUCIsImVuYyI6IkEyNTZHQ00ifQ"                                # A.1.1
assert unb64u(rsa_hdr) == b'{"alg":"RSA-OAEP","enc":"A256GCM"}'
out = aesgcm.encrypt(iv, plaintext, rsa_hdr.encode("ascii"))
assert out[:-16] == rfc_ct and out[-16:] == rfc_tag       # ciphertext AND tag match the RFC

# 1. Same key, same IV, same plaintext — A.1's header with only the alg value swapped.
dir_hdr = b64u(b'{"alg":"dir","enc":"A256GCM"}')
out2 = aesgcm.encrypt(iv, plaintext, dir_hdr.encode("ascii"))
ct, tag = out2[:-16], out2[-16:]
assert ct == rfc_ct                     # unchanged: CTR keystream does not depend on AAD
assert tag != rfc_tag                   # changed:   GHASH does

jwe      = f"{dir_hdr}..{b64u(iv)}.{b64u(ct)}.{b64u(tag)}"
bad_tag  = tag[:-1] + bytes([tag[-1] ^ 1])
jwe_tag  = f"{dir_hdr}..{b64u(iv)}.{b64u(ct)}.{b64u(bad_tag)}"
kid_hdr  = b64u(b'{"alg":"dir","enc":"A256GCM","kid":"kat"}')
jwe_aad  = f"{kid_hdr}..{b64u(iv)}.{b64u(ct)}.{b64u(tag)}"   # same iv/ct/tag, altered AAD

assert aesgcm.decrypt(iv, ct + tag, dir_hdr.encode("ascii")) == plaintext
for label, (c, a) in {"tag": (ct + bad_tag, dir_hdr), "aad": (ct + tag, kid_hdr)}.items():
    try:
        aesgcm.decrypt(iv, c, a.encode("ascii")); print("BUG:", label, "accepted")
    except InvalidTag:
        print(label, "correctly rejected")

print("KAT_VALID_JWE        =", jwe)
print("KAT_TAMPERED_TAG_JWE  =", jwe_tag)
print("KAT_TAMPERED_AAD_JWE  =", jwe_aad)

File: hsm-worker/src/infrastructure/adapters/outgoing/self_test_probes/crypto_a256gcm_kat.rs
// SPDX-FileCopyrightText: 2026 Digg - Agency for Digital Government
//
// SPDX-License-Identifier: EUPL-1.2

//! FPT_TST.1 Application Note 30 — "correct operation of ... cryptographic libraries".
//!
//! Known-answer test for `dir` + A256GCM content decryption, the algorithm protecting
//! every post-authentication inner JWE. Runs through the unmodified
//! `hsm_common::jose::jwe_decrypt_dir`.
//!
//! No published `dir`+A256GCM compact-JWE vector exists — RFC 7520 §5.6's `dir` example is
//! A128GCM, and the A256GCM examples (RFC 7516 §A.1, RFC 7520 §5.2) wrap the CEK with
//! RSA-OAEP. This vector is therefore built from RFC 7516 §A.1's published CEK, IV and
//! plaintext under a `dir` protected header, with the authentication tag computed by an
//! implementation independent of this codebase (Python `cryptography`). Because AES-GCM's
//! CTR keystream depends only on key and IV, the ciphertext below is byte-for-byte the
//! ciphertext RFC 7516 §A.1.6 publishes; only the 16-byte tag differs, the AAD having
//! changed with the header.
//!
//! Two negative cases, not one. An AEAD fails open by returning plaintext without
//! checking the tag, and the two authenticated regions fail independently: a tampered tag
//! catches a missing GHASH comparison, and a tampered protected header catches a tag
//! computed over the wrong authenticated data — a bug the tag case passes straight over.

use hsm_common::jose;

use crate::application::self_test_spi_port::{SelfTestError, SelfTestProbe, TsfClaim};

// RFC 7516 §A.1.2 — the 256-bit Content Encryption Key, verbatim.
pub(crate) const KAT_SESSION_KEY: [u8; 32] = [
    177, 161, 244, 128, 84, 143, 225, 115, 63, 180, 3, 255, 107, 154, 212, 246, 138, 7, 110, 91,
    112, 46, 34, 105, 47, 130, 203, 46, 122, 234, 64, 252,
];

// Protected header {"alg":"dir","enc":"A256GCM"} (RFC 7516 §A.1.1's header with only the
// alg value swapped), the JWE Encrypted Key empty as `dir` requires, IV from §A.1.4 and
// ciphertext from §A.1.6 unchanged, tag recomputed for the new AAD.
pub(crate) const KAT_VALID_JWE: &str = "eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..48V1_ALb6US04U3b.5eym8TW_c8SuK0ltJ3rpYIzOeDQz7TALvtu6UG9oMo4vpzs9tX_EFShS8iB7j6jiSdiwkIr3ajwQzaBtQD_A.Q_HtEPSRG-ujQXQqWM1jlg";

// Low bit of the tag's last byte flipped (g -> w). Header, IV and ciphertext identical to
// the valid JWE. Must be rejected.
pub(crate) const KAT_TAMPERED_TAG_JWE: &str = "eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..48V1_ALb6US04U3b.5eym8TW_c8SuK0ltJ3rpYIzOeDQz7TALvtu6UG9oMo4vpzs9tX_EFShS8iB7j6jiSdiwkIr3ajwQzaBtQD_A.Q_HtEPSRG-ujQXQqWM1jlw";

// Protected header extended with `"kid":"kat"`; IV, ciphertext and tag identical to the
// valid JWE. The header is the AAD, so this must be rejected on the tag comparison.
pub(crate) const KAT_TAMPERED_AAD_JWE: &str = "eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwia2lkIjoia2F0In0..48V1_ALb6US04U3b.5eym8TW_c8SuK0ltJ3rpYIzOeDQz7TALvtu6UG9oMo4vpzs9tX_EFShS8iB7j6jiSdiwkIr3ajwQzaBtQD_A.Q_HtEPSRG-ujQXQqWM1jlg";

// RFC 7516 §A.1 — the plaintext, exactly.
pub(crate) const KAT_EXPECTED_PLAINTEXT: &[u8] =
    b"The true sign of intelligence is not knowledge but imagination.";

pub struct CryptoA256GcmKatProbe;

impl SelfTestProbe for CryptoA256GcmKatProbe {
    fn name(&self) -> &'static str {
        "crypto_a256gcm_kat"
    }

    fn claim(&self) -> TsfClaim {
        TsfClaim::CryptographicLibraries
    }

    fn probe(&self) -> Result<(), SelfTestError> {
        let plaintext =
            jose::jwe_decrypt_dir(KAT_VALID_JWE, &KAT_SESSION_KEY).map_err(|_| SelfTestError {
                detail: "A256GCM KAT: pinned positive vector failed to decrypt".to_string(),
            })?;
        if plaintext != KAT_EXPECTED_PLAINTEXT {
            return Err(SelfTestError {
                detail: "A256GCM KAT: positive vector decrypted but plaintext did not match"
                    .to_string(),
            });
        }

        if jose::jwe_decrypt_dir(KAT_TAMPERED_TAG_JWE, &KAT_SESSION_KEY).is_ok() {
            return Err(SelfTestError {
                detail: "A256GCM KAT: tampered authentication tag was incorrectly accepted"
                    .to_string(),
            });
        }

        if jose::jwe_decrypt_dir(KAT_TAMPERED_AAD_JWE, &KAT_SESSION_KEY).is_ok() {
            return Err(SelfTestError {
                detail: "A256GCM KAT: tampered protected header was incorrectly accepted"
                    .to_string(),
            });
        }

        Ok(())
    }
}
